framework/resource/retryresource/resource.go

```diff
@@ -43,188 +33,54 @@ func New(config Config) (*Resource, error) {
 		config.BackOff = backoff.NewExponentialBackOff()
 	}
 
-	r := &Resource{
-		logger: config.Logger.With(
-			"underlyingResource", config.Resource.Underlying().Name(),
-		),
-		resource: config.Resource,
-
-		backOff: config.BackOff,
-	}
-
-	return r, nil
-}
-
-func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
-	var err error
+	// If the resource is instance of *framework.CRUDResource we wrap it
+	// with a crudResource otherwise we wrap with basicResource.
+	crudResource, ok := config.Resource.Underlying().(*framework.CRUDResource)
+	if ok {
+		c := crudResourceConfig{
+			Logger:   config.Logger,
+			Resource: crudResource.CRUDResourceOps,
 
-	var v interface{}
-	o := func() error {
-		v, err = r.resource.GetCurrentState(ctx, obj)
-		if err != nil {
-			return microerror.Mask(err)
+			BackOff: config.BackOff,
 		}
 
-		return nil
-	}
-
-	n := func(err error, dur time.Duration) {
-		r.logger.LogCtx(ctx, "warning", fmt.Sprintf("retrying 'GetCurrentState' due to error (%s)", err.Error()))
-	}
-
-	err = backoff.RetryNotify(o, r.backOff, n)
-	if err != nil {
-		return nil, microerror.Mask(err)
-	}
-
-	return v, nil
-}
-
-func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
-	var err error
-
-	var v interface{}
-	o := func() error {
-		v, err = r.resource.GetDesiredState(ctx, obj)
+		wrappedOps, err := newCRUDResource(c)
 		if err != nil {
-			return microerror.Mask(err)
+			return nil, microerror.Mask(err)
 		}
 
-		return nil
-	}
+		crudResource.CRUDResourceOps = wrappedOps
 
-	n := func(err error, dur time.Duration) {
-		r.logger.LogCtx(ctx, "warning", fmt.Sprintf("retrying 'GetDesiredState' due to error (%s)", err.Error()))
+		return crudResource, nil
+	} else {
+		// TODO uncomment when new Resource interface is created.
+		//c := basicResourceConfig{
+		//	Logger:   config.Logger,
+		//	Resource: config.Resource,
+		//
+		//	BackOff: config.BackOff,
+		//}
+		//
+		//r, err := newBasicResource(c)
+		//if err != nil {
+		//	return nil, microerror.Mask(err)
+		//}
+		//
+		//return r, nil
 	}
```

Comment:
Diff is a mess. It's better to look at the whole file: https://github.com/giantswarm/operatorkit/blob/cb570a44556cac8cfee3e3d848721fdb8cdc2b1f/framework/resource/retryresource/resource.go

The idea here is wrap `Resource.CRUDResourceOps` with `crudResource` if resource is instance of `*framework.CRUDResource` (`crudResource, ok := config.Resource.Underlying().(*framework.CRUDResource)`) otherwise wrap `Resource` with `basicRessource`. That way we can still use `Wrap` method to wrap mixed types of resources.

Reply:
When we assert interfaces here at runtime we should have a default matching to handle all unknown error cases.

Reply:
There aren't error cases. It's either `*framework.CRUDResource` or `framework.Resource`.

Note that `*framework.CRUDResource` is also `framework.Resource`, there's just a special case for it.

Reply:
We are type casting an interface. I think there can just be anything and in case we do not receive one of the two expected interfaces, we should handle this error case to be safe.